### PR TITLE
[Driver] Add -driver-filelist-threshold to set the number of inputs beyond which filelists are used

### PR DESCRIPTION
**Pull request by nathawes**
When generating a compiler invocation in `driver::createCompilerInvocation()` we end up using filelists if the number of inputs is > 128 (to work around command line arg limits). We never actually write them out though, and so fail when parsing the frontend arguments that reference them.

As this function is called frequently by SourceKit and command line limits aren't a concern here, this patch makes the 128 threshold value configurable via a new `-driver-filelist-threshold` option. This is set to its maximum value in `driver::createCompilerInvocation()` to ensure filelists aren't used. 

This new option makes the existing `-driver-use-filelists` flag (that forces filelists to be used) redundant as it's now equivalent to `-driver-filelist-threshold=0`.

Resolves rdar://problem/38231888

### CODE REVIEW

**Review by graydon — CHANGES_REQUESTED** [review]
I understand where this is coming from and recognize that you need to do something, but feel a little uneasy about having yet another factor contributing to the file-list-use decision. I suspect @davidungar will also feel that the driver ought to warn the user when it is passed -driver-use-filelists and then it chooses (due to this flag being set on an internal operation) to ignore the argument. Best to run this by him as well, maybe he has an idea on how to factor or consolidate the logic a little more tightly.

**Review by davidungar — CHANGES_REQUESTED** [review]
The driver is already creaking and groaning a bit with special cases as a result of a fine original design being incrementally complexified through the years, AFAICT.
This is nobody's fault, but the result is that it becomes extremely hard to modify things without creating bugs, because of duplicated functionality at various levels of abstraction.
This solution to the problem seems like another special case. Why not get createCompilerInvocation to write out the file list by calling the same code that it normally does? Is there some reason why that cannot work? Thanks.

**Comment by nathawes** [comment]
@davidungar we could write the file lists out in createCompilerInvocation but then SourceKit (it's only caller) would need to manage their lifetime, which I was hoping to avoid, especially since we don't care about the number of inputs. I of course don't want to make the driver more difficult to maintain either though!

Would you be more ok with making the below in Toolchain.h configurable?
```
/// The limit for passing a list of files on the command line.
    static const size_t TOO_MANY_FILES = 128;
```
That way SourceKit can indicate it doesn't care about the limit and the logic in the various `shouldUse*FileList*()` methods wouldn't need to change (`-driver-use-filelists` isn't really a problem – `createCompilerInvocation` can filter it out if need be).

**Comment by davidungar** [comment]
@nathawes Thanks for sharing our concern for driver maintainability!
Yes, configuring TOO_MANY_FILES would be better, but I'm still quite uneasy about that approach. For instance, it could bite us if we ever go to a pure-file-list invocation convention. (That will likely never happen, but it is an example of the coupling that this approach creates.) The way the SourceKit goes through the driver and then reuses the command-line arguments feels like something I could easily break by accident as I evolve the driver. This file-list issue is an instance of that, as you know all too well:) 
What do you do about cleaning up other temporary files from SourceKit?

Here's another approach...I'm just putting it up there:
Take the code in `Compilation::performJobs` that follows:
    if (!SaveTemps) {
        for (const auto &pathPair : TempFilePaths) {
          if (!abnormalExit || pathPair.getValue() == PreserveOnSignal::No)
            (void)llvm::sys::fs::remove(pathPair.getKey());
        }
      }
and factor it out into a public member function of Compilation, called something like `cleanUpTempFiles`.  Call it from both `performJobs` and somewhere in source kit, or maybe even call it from the destructor for Compilation instead. 
Anyway, something like this. Does this make any sense at all? I realize it's more work, but in the long run it has the benefit of not increasing the coupling. What do you think?

**Comment by jrose-apple** [comment]
It probably makes sense for SourceKit not to hit the disk at all here; the point of `driver::createCompilerInvocation` is to immediately, in-process, give you a SingeCompile frontend CompilerInvocation in the same way that the driver would. That said, producing a very long argument list may not be a good answer either…

**Comment by davidungar** [comment]
@jrose-apple Yes, I thought about the SourceKit-not-hitting-the-disk argument, but then I thought about how the compiler hits the disk in so many other ways. (Some?) Inputs can be supplied by buffers, but AFAICT, the basic design is about reading and writing files, although we may want to get away from that in the future. (As far as I know, hitting-the-disk is really just hitting the file-system cache in the OS, maybe.)

It feels like the way that SourceKit reuses the driver functionality could be a better architecture. As it is, the driver code thinks it is preparing commands to be run as separate processes, and the SourceKit tries to tweak it into not running separate processes. The SourceKit scenario goes through all the logic in the driver to encode the task into command-line arguments, then goes through all the logic in the frontend to translate the command-line arguments into a frontend data structure. Sigh...

Meanwhile, the driver has a constant, TOO_MANY_FILES, and a parameter to force the use of file lists.
Thinking about it now, I suppose my favored short-term solution is to *change* the command-line arguments to the driver by:
1. eliminating -use-file-maps, and 
2. introducing -file-map-threshold <n>

Setting the threshold to zero would replace the functionality of -use-file-maps, setting it to infinity (or a large number) would be like -never-use-file-maps. Passing too_many_files down through the driver architecture would satisfy source kit, though it leaves some fragility in place. IMO, in the long term, we'll want to redesign the way source kit reuses the compiler.

@jrose-apple What do you think? Is this reasonable? Can we make this change?

**Comment by jrose-apple** [comment]
I disagree that SourceKit is doing the wrong thing, but that's separable. The approach you describe is reasonable. Driver options are forever, though, so we'll have to keep supporting the old flag by mapping it to the new one (possibly with a warning). @graydon, @nathawes?

**Comment by jrose-apple** [comment]
Oh, and I was also worried about having to clean up temporary files at some undetermined point in the future, but it's *not* some undetermined point in the future; it's later in `driver::createCompilerInvocation`. So that concern at least is less relevant.

**Comment by nathawes** [comment]
Oh, so it is! Looking at my original diagnosis in the radar I even saw that at some point, so sorry all for the misleading argument.

In that case I'm ok with changing to write out and clean up the file lists in createCompilerInvocation. It does seem non-ideal in terms of disk access, but as @davidungar says maybe it doesn't make that much difference in practice? @akyrtzi what do you think? Should we just try and measure?

**Comment by nathawes** [comment]
Spoke to @akyrtzi in person and he shares the IO concern so I'll go ahead with the `-file-map-threshold` approach.

**Comment by davidungar** [comment]
I'd push back on the IO concern, given that filemaps only get written for large compile jobs anyway. But since we need this fixed urgently (that's my understanding), I would vote for landing with the file-map-threshold approach, if that's the quickest way for you to get there. I'd love to understand more about your and @akyrtzi 's IO concerns. You're probably seeing something I'm not. At any rate, feel free to go ahead with `-file-map-threshold`.

**Comment by akyrtzi** [comment]
>  I was also worried about having to clean up temporary files at some undetermined point in the future, but it's not some undetermined point in the future; it's later in driver::createCompilerInvocation

To make sure I understood, what would happen is that a temporary file would be created, that file would be read so that its contents fill out the invocation structure, and then the file gets disposed ? Is this correct ?

**Comment by akyrtzi** [comment]
> Only in the case where there are enough input files will any files be created and read.

When the project is setup with many files, then every time we call `driver::createCompilerInvocation` a temporary file will be created. And we'll call it *a lot* for an open document, for example every time code-completion is invoked which can happen multiple times while the user is typing (due to automatic code-completion functionality), along with additional calls for other semantic functionality.

> In case I'm wrong, I just did a quick test: time to create, open, write, close, open read close remove a 1K file was 45 ms.

Was this on a spinning disk or SSD ? Was there I/O contention in the system or not ?
These things just add up and we should not be so "laissez faire" about creating temporary files in functions that would not be expected that they hit the disk at all.

The function `createCompilerInvocation` taking 45 ms (even if not all the time, but in some specific cases, which can be worse due to its unpredictability) is very unintuitive and I personally find it unacceptable. I believe we don't use it for syntactic parsing but if for some reason we needed to use it for that then we would have a *massive* performance regression which would probably go unnoticed initially because it would hit only on some specific project setups and not all the time. To give you more context, something like an over 20~30ms regression in syntactic parsing would be noticeable by users in lag while they are typing.

**Comment by davidungar** [comment]
Great! I’m convinced. In that case the most elegant solution IMO is probably the threshold setting plus a constant called something like NeverUseFileList = to the largest possible unsigned int. And a comment there explaining in essence why it’s there so we don’t break source kit (again) in the future. And an assertion in the file list code that the threshold is not Never. 

But it’s ok if the above is too much. 

- David

**Comment by nathawes** [comment]
@swift-ci please test

**Comment by nathawes** [comment]
I updated the patch to add -driver-filelist-threshold and to map -driver-use-filelists to -driver-filelist-threshold=0 and warn when we see it.
@davidungar would you mind taking another look over the latest changes?

**Review by davidungar — APPROVED** [review]
LGTM! Just one little nitpick on the help text for the new option. Tx.

File: include/swift/Option/Options.td [hunk 1]
```diff
@@ -102,6 +102,11 @@ def driver_show_job_lifecycle : Flag<["-"], "driver-show-job-lifecycle">,
   HelpText<"Show every step in the lifecycle of driver jobs">;
 def driver_use_filelists : Flag<["-"], "driver-use-filelists">,
   InternalDebugOpt, HelpText<"Pass input files as filelists whenever possible">;
+def driver_filelist_threshold : Separate<["-"], "driver-filelist-threshold">,
+  InternalDebugOpt, HelpText<"Pass input files as filelists if there are more than <n>">,
```
[the remark this answers is not in the document]
Comment (jrose-apple) [reply include/swift/Option/Options.td]: Looks like the old comment never got updated either.

**Comment by benlangmuir** [comment]
The failure:

```
Could not open input file '/...../Driver/Output/filelist-threshold.swift.tmp/out.txt': No such file or directory
```

Is because the test passed `--input-file` to FileCheck when I think it intended to pipe the output; You wouldn't see this if you ran the test previously because the file would already exist from a previous run.

**Comment by benlangmuir** [comment]
@swift-ci please smoke test

**Comment by benlangmuir** [comment]
It turns out this issue was much worse than it looked, because it affects outputs as well as inputs.  This means that even with relatively small invocations of 5-6 files would fail in sourcekitd.  For example, code-completion with 5 input files was failing.

**Comment by akyrtzi** [comment]
We need to get these changes ASAP to address rdar://39398871, if you have any additional feedback please provide it post-commit.

**Pull request merged**